File: demos/api/import_perf.py
import gmsh
import math

gmsh.initialize()

N = 2500

tic = gmsh.logger.getWallTime()
try:
    import numpy
    # x, y, z coordinates of all the nodes
    coords = numpy.empty((N + 1) * (N + 1) * 3)
    # tags of corresponding nodes
    nodes = numpy.empty((N + 1) * (N + 1))
    # connectivities (node tags) of triangle elements
    tris = numpy.empty(N * N * 2 * 3)
except:
    coords = [0] * (N + 1) * (N + 1) * 3
    nodes = [0] * (N + 1) * (N + 1)
    tris = [0] * N * N * 2 * 3

def tag(i, j):
    return (N + 1) * i + j + 1

k = 0
l = 0
for i in range(N + 1):
    for j in range(N + 1):
        nodes[k] = tag(i, j)
        coords[3 * k] = float(i) / N
        coords[3 * k + 1] = float(j) / N
        coords[3 * k + 2] = 0.05 * math.sin(10 * float(i + j) / N)
        k = k + 1
        if i > 0 and j > 0:
            tris[6 * l] = tag(i - 1, j - 1)
            tris[6 * l + 1] = tag(i, j - 1)
            tris[6 * l + 2] = tag(i - 1, j)
            tris[6 * l + 3] = tag(i, j - 1)
            tris[6 * l + 4] = tag(i, j)
            tris[6 * l + 5] = tag(i - 1, j)
            l = l + 1

toc = gmsh.logger.getWallTime()
print("==> created nodes and connectivities in {} seconds".format(toc - tic))

tic = gmsh.logger.getWallTime()
surf = gmsh.model.addDiscreteEntity(2)
toc = gmsh.logger.getWallTime()
print("==> created surface in {} seconds".format(toc - tic))

tic = gmsh.logger.getWallTime()
gmsh.model.mesh.addNodes(2, 1, nodes, coords)
toc = gmsh.logger.getWallTime()
print("==> imported nodes in {} seconds".format(toc - tic))

tic = gmsh.logger.getWallTime()
gmsh.model.mesh.addElementsByType(1, 2, [], tris)
toc = gmsh.logger.getWallTime()
print("==> imported elements in {} seconds".format(toc - tic))

tic = gmsh.logger.getWallTime()
gmsh.option.setNumber("Mesh.Binary", 1)
gmsh.write("import_perf.msh")
toc = gmsh.logger.getWallTime()
print("==> wrote to disk in {} seconds".format(toc - tic))

tic = gmsh.logger.getWallTime()
gmsh.merge("import_perf.msh")
toc = gmsh.logger.getWallTime()
print("==> read from disk in {} seconds".format(toc - tic))

#gmsh.fltk.run()

gmsh.finalize()

File: demos/api/import_perf.cpp
#include <gmsh.h>
#include <math.h>

int main()
{
  gmsh::initialize();

  int N = 2500;

  double tic = gmsh::logger::getWallTime();
  // x, y, z coordinates of all the nodes
  std::vector<double> coords((N + 1) * (N + 1) * 3);
  // tags of corresponding nodes
  std::vector<std::size_t> nodes((N + 1) * (N + 1));
  // connectivities (node tags) of triangle elements
  std::vector<std::size_t> tris(N * N * 2 * 3);

  auto tag = [N](int i, int j) {
    return (N + 1) * i + j + 1;
  };

  int k = 0, l = 0;
  for(int i = 0; i < N + 1; i++) {
    for(int j = 0; j < N + 1; j++) {
      nodes[k] = tag(i, j);
      coords[3 * k] = (double)i / N;
      coords[3 * k + 1] = (double)j / N;
      coords[3 * k + 2] = 0.05 * sin(10 * (double)(i + j) / N);
      k++;
      if(i > 0 && j > 0) {
        tris[6 * l] = tag(i - 1, j - 1);
        tris[6 * l + 1] = tag(i, j - 1);
        tris[6 * l + 2] = tag(i - 1, j);
        tris[6 * l + 3] = tag(i, j - 1);
        tris[6 * l + 4] = tag(i, j);
        tris[6 * l + 5] = tag(i - 1, j);
        l++;
      }
    }
  }

  double toc = gmsh::logger::getWallTime();
  printf("==> created nodes and connectivities in %g seconds\n", toc - tic);

  tic = gmsh::logger::getWallTime();
  int surf = gmsh::model::addDiscreteEntity(2);
  toc = gmsh::logger::getWallTime();
  printf("==> created surface in %g seconds\n", toc - tic);

  tic = gmsh::logger::getWallTime();
  gmsh::model::mesh::addNodes(2, 1, nodes, coords);
  toc = gmsh::logger::getWallTime();
  printf("==> imported nodes in %g seconds\n", toc - tic);

  tic = gmsh::logger::getWallTime();
  gmsh::model::mesh::addElementsByType(1, 2, {}, tris);
  toc = gmsh::logger::getWallTime();
  printf("==> imported elements in %g seconds\n", toc - tic);

  tic = gmsh::logger::getWallTime();
  gmsh::option::setNumber("Mesh.Binary", 1);
  gmsh::write("import_perf.msh");
  toc = gmsh::logger::getWallTime();
  printf("==> wrote to disk in %g seconds\n", toc - tic);

  tic = gmsh::logger::getWallTime();
  gmsh::merge("import_perf.msh");
  toc = gmsh::logger::getWallTime();
  printf("==> read from disk in %g seconds\n", toc - tic);

  //gmsh::fltk::run();

  gmsh::finalize();
  return 0;
}
